File: register-alloc-sim/src/core/linearScan.ts
// ==========================================
// Poletto-Sarkar Linear Scan Register Allocation
// ==========================================

import type { TACInstruction, LiveInterval, AllocationResult, SimulationStep } from './types';

// ---- Step 1: Compute live intervals ----
// interval[v] = [first definition index, last use index]
export function computeLiveIntervals(instructions: TACInstruction[]): LiveInterval[] {
  const intervalMap = new Map<string, { start: number; end: number }>();

  function isVar(s: string): boolean {
    return /^[a-zA-Z_]\w*$/.test(s);
  }

  function touch(v: string, idx: number): void {
    if (!isVar(v)) return;
    const existing = intervalMap.get(v);
    if (!existing) {
      intervalMap.set(v, { start: idx, end: idx });
    } else {
      existing.start = Math.min(existing.start, idx);
      existing.end = Math.max(existing.end, idx);
    }
  }

  for (let i = 0; i < instructions.length; i++) {
    const instr = instructions[i];
    if (instr.result) touch(instr.result, i);
    if (instr.arg1) touch(instr.arg1, i);
    if (instr.arg2) touch(instr.arg2, i);
  }

  // Convert to LiveInterval array
  const intervals: LiveInterval[] = [];
  for (const [variable, { start, end }] of intervalMap) {
    intervals.push({ variable, start, end, spilled: false });
  }

  return intervals;
}

// ---- Step 2: Linear Scan ----
export function linearScanAllocation(
  instructions: TACInstruction[],
  k: number
): AllocationResult {
  const steps: SimulationStep[] = [];
  let stepIdx = 0;

  const intervals = computeLiveIntervals(instructions);

  if (intervals.length === 0) {
    return { assignments: {}, spills: [], spillCode: [], steps, registersUsed: 0, totalSpills: 0 };
  }

  // Sort by start point
  intervals.sort((a, b) => a.start - b.start);

  steps.push({
    stageId: 'linearScan',
    stepIndex: stepIdx++,
    description: `Computed ${intervals.length} live intervals, k=${k} registers`,
    detail: 'Sorted intervals by start point. Beginning linear scan.',
    activeIntervals: [],
    assignmentsSoFar: {},
  });

  // Active list: intervals currently holding a register (sorted by end point)
  let active: LiveInterval[] = [];
  const freeRegisters: Set<number> = new Set(Array.from({ length: k }, (_, i) => i));
  const assignments: Record<string, number> = {};
  const spills: string[] = [];

  const getActiveNames = () => active.map((a) => a.variable);

  for (const interval of intervals) {
    // --- Expire old intervals ---
    const expired: LiveInterval[] = [];
    const stillActive: LiveInterval[] = [];
    for (const a of active) {
      if (a.end < interval.start) {
        expired.push(a);
      } else {
        stillActive.push(a);
      }
    }

    for (const exp of expired) {
      freeRegisters.add(exp.register!);
    }
    active = stillActive;

    if (expired.length > 0) {
      steps.push({
        stageId: 'linearScan',
        stepIndex: stepIdx++,
        description: `Expired: [${expired.map((e) => e.variable).join(', ')}] — registers freed`,
        detail: `Intervals ending before position ${interval.start} are freed. Free registers: {${[...freeRegisters].map((r) => `R${r}`).join(', ')}}`,
        activeIntervals: getActiveNames(),
        assignmentsSoFar: { ...assignments },
      });
    }

    // --- Allocate or spill ---
    if (freeRegisters.size === 0) {
      // Spill: choose the interval with the farthest end point
      active.sort((a, b) => b.end - a.end);
      const spillCandidate = active[0];

      if (spillCandidate && spillCandidate.end > interval.end) {
        // Spill the farthest-ending interval, give its register to current
        const reg = spillCandidate.register!;
        spillCandidate.spilled = true;
        delete assignments[spillCandidate.variable];
        spills.push(spillCandidate.variable);
        active.splice(0, 1); // remove it

        interval.register = reg;
        interval.spilled = false;
        assignments[interval.variable] = reg;
        active.push(interval);
        active.sort((a, b) => a.end - b.end);

        steps.push({
          stageId: 'linearScan',
          stepIndex: stepIdx++,
          description: `Spill: "${spillCandidate.variable}" (ends at ${spillCandidate.end}) — give R${reg} to "${interval.variable}"`,
          detail: `"${spillCandidate.variable}" has a farther endpoint, so it's spilled to memory.`,
          activeIntervals: getActiveNames(),
          spilledVariables: [...spills],
          assignmentsSoFar: { ...assignments },
        });
      } else {
        // Spill current interval
        interval.spilled = true;
        spills.push(interval.variable);

        steps.push({
          stageId: 'linearScan',
          stepIndex: stepIdx++,
          description: `Spill: "${interval.variable}" — no free registers and all active end later`,
          detail: `"${interval.variable}" is spilled to memory.`,
          activeIntervals: getActiveNames(),
          spilledVariables: [...spills],
          assignmentsSoFar: { ...assignments },
        });
      }
    } else {
      // Assign the lowest-numbered free register
      const reg = Math.min(...freeRegisters);
      freeRegisters.delete(reg);
      interval.register = reg;
      interval.spilled = false;
      assignments[interval.variable] = reg;
      active.push(interval);
      active.sort((a, b) => a.end - b.end);

      steps.push({
        stageId: 'linearScan',
        stepIndex: stepIdx++,
        description: `Assigned R${reg} to "${interval.variable}" [${interval.start}–${interval.end}]`,
        detail: `Active set: [${getActiveNames().join(', ')}]. Free registers after: {${[...freeRegisters].map((r) => `R${r}`).join(', ')}}`,
        activeIntervals: getActiveNames(),
        assignmentsSoFar: { ...assignments },
      });
    }
  }

  const registersUsed = new Set(Object.values(assignments)).size;

  steps.push({
    stageId: 'linearScan',
    stepIndex: stepIdx++,
    description: `Linear scan complete! ${registersUsed} registers used, ${spills.length} spills`,
    detail: spills.length === 0
      ? 'All variables successfully allocated.'
      : `Spilled: ${spills.join(', ')}`,
    assignmentsSoFar: { ...assignments },
    spilledVariables: [...spills],
  });

  return {
    assignments,
    spills,
    spillCode: [],
    steps,
    registersUsed,
    totalSpills: spills.length,
    // also expose intervals for the Gantt chart
  } as AllocationResult & { intervals: LiveInterval[] };
}

// Re-export for convenience
export { computeLiveIntervals as getLiveIntervals };

File: register-alloc-sim-v2/backend/linear_scan.py
"""
Custom Linear Scan register allocator — Poletto-Sarkar algorithm.

Operates on live intervals derived from the LLVM IR's liveness data.
Returns a register assignment dict, spill list, and Gantt-chart data.
"""
from typing import Dict, List, Set, Optional, Tuple
from dataclasses import dataclass, field
import re


@dataclass
class LiveInterval:
    variable: str
    start: int      # instruction index of first definition
    end: int        # instruction index of last use
    register: Optional[str] = None
    spilled: bool = False


def compute_live_intervals(blocks: list) -> List[LiveInterval]:
    """
    Compute linear live intervals from a list of IRBasicBlock objects.

    We assign a monotonically increasing instruction index across all blocks
    (flattening the CFG to a linear order) and then compute [start, end]
    for each SSA value.
    """
    _RESULT_RE = re.compile(r"^\s*(%[\w.]+)\s*=")
    _USE_RE = re.compile(r"%[\w.]+")

    instr_index = 0
    first_def: Dict[str, int] = {}
    last_use: Dict[str, int] = {}

    for block in blocks:
        for instr in block.instructions:
            # Definition
            m = _RESULT_RE.match(instr)
            if m:
                var = m.group(1)
                if var not in first_def:
                    first_def[var] = instr_index

            # All uses
            refs = _USE_RE.findall(instr)
            for ref in refs:
                last_use[ref] = instr_index

            instr_index += 1

    # Build intervals — a variable's interval spans from first def to last use
    intervals: List[LiveInterval] = []
    all_vars = set(first_def) | set(last_use)
    for var in sorted(all_vars):
        start = first_def.get(var, last_use.get(var, 0))
        end = last_use.get(var, start)
        intervals.append(LiveInterval(variable=var, start=start, end=end))

    return sorted(intervals, key=lambda iv: iv.start)


def linear_scan(
    intervals: List[LiveInterval],
    k: int = 6,
) -> Dict:
    """
    Poletto-Sarkar Linear Scan register allocator.

    Returns
    -------
    {
        "registerMap":   { variable: register_name | "spilled" },
        "spills":        [ variable, ... ],
        "spillCount":    int,
        "registerCount": int,
        "intervals":     [ { variable, start, end, register, spilled } ],
        "steps":         [ { description, activeList, assigned } ]
    }
    """
    register_pool = [f"r{i}" for i in range(k)]
    free_regs: List[str] = list(register_pool)
    active: List[LiveInterval] = []   # sorted by end point
    spills: List[str] = []
    steps: List[Dict] = []

    def record(desc: str):
        steps.append({
            "description": desc,
            "activeList": [iv.variable for iv in active],
            "assigned": {iv.variable: iv.register for iv in active if iv.register},
        })

    for iv in intervals:
        # Expire old intervals
        expired = [a for a in active if a.end < iv.start]
        for e in expired:
            active.remove(e)
            if e.register:
                free_regs.append(e.register)
            record(f"Expire {e.variable} (end={e.end} < start={iv.start})")

        if len(active) == k:
            # Spill: choose the interval with the farthest endpoint
            spill_target = max(active, key=lambda a: a.end)
            if spill_target.end > iv.end:
                iv.register = spill_target.register
                spill_target.register = None
                spill_target.spilled = True
                active.remove(spill_target)
                active.append(iv)
                spills.append(spill_target.variable)
                record(f"Spill {spill_target.variable} (farthest end) → give r to {iv.variable}")
            else:
                iv.spilled = True
                spills.append(iv.variable)
                record(f"Spill {iv.variable} (its end ≤ current farthest active)")
        else:
            if free_regs:
                iv.register = free_regs.pop(0)
            else:
                iv.spilled = True
                spills.append(iv.variable)
                record(f"No free registers — spill {iv.variable}")
                continue
            active.append(iv)
            active.sort(key=lambda a: a.end)
            record(f"Assign {iv.variable} → {iv.register}")

    register_map = {
        iv.variable: iv.register if iv.register else "spilled"
        for iv in intervals
    }
    registers_used = len({iv.register for iv in intervals if iv.register})

    return {
        "registerMap": register_map,
        "spills": spills,
        "spillCount": len(spills),
        "registerCount": registers_used,
        "intervals": [
            {
                "variable": iv.variable,
                "start": iv.start,
                "end": iv.end,
                "register": iv.register,
                "spilled": iv.spilled,
            }
            for iv in intervals
        ],
        "steps": steps,
    }
